### PR TITLE
feat: add typehint on adapter

### PR DESCRIPTION
**Pull request by KevinArtus**

### CODE REVIEW

File: src/Gaufrette/Adapter/AwsS3.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/AwsS3.php:176]:
```suggestion
    public function mimeType(string $key): string|bool
```

File: src/Gaufrette/Adapter/AwsS3.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/AwsS3.php:318]: not available on 8.0.
```suggestion
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:146]: same, why not `string $content` ?

File: src/Gaufrette/Adapter/Flysystem.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/Flysystem.php:19]:
```suggestion
    public function __construct(AdapterInterface $adapter, ?Config $config = null)
```

Comment (Zusoy) [inline src/Gaufrette/Adapter/Flysystem.php:19]:
```suggestion
    public function __construct(AdapterInterface $adapter, ?Config $config = null)
```

File: src/Gaufrette/Adapter/PhpseclibSftp.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/PhpseclibSftp.php:62]: Same, wy not `string $content` ?

File: src/Gaufrette/Adapter/PhpseclibSftp.php
Comment (Zusoy) [inline src/Gaufrette/Adapter/PhpseclibSftp.php:191]:
```suggestion
    protected function ensureDirectoryExists(string $directory, bool $create): void
```

File: src/Gaufrette/Adapter.php
Comment (KevinArtus) [reply src/Gaufrette/Adapter.php:25]: This class is an interface implemented by lot of classes.
Those class implements write which will call methods where content is used as a resource

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (KevinArtus) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:54]: Because isset will also test if the data is not null. Not `array_key_exists`

File: src/Gaufrette/Adapter/AwsS3.php
Comment (KevinArtus) [reply src/Gaufrette/Adapter/AwsS3.php:51]: same

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (muchafm) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:23]: @KevinArtus we said that we will add the property promotion et readonly once the typehint will be merge right ?

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (muchafm) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:307]:
```suggestion
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Adapter/AwsS3.php
Comment (muchafm) [inline src/Gaufrette/Adapter/AwsS3.php:23]: same here

**Review by muchafm — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (PedroTroller) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:26]: Not the goal of these PR, will be done in an other.

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (PedroTroller) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:57]: And... it's not the goal of there PR :kissing:

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter.php
Comment (KevinArtus) [reply src/Gaufrette/Adapter.php:25]: For example, the method `write` on `Gaufrette\Adapter\AzureBlobStorage` calls the foo method `guessContentType`.

This needs a resource to execute its code

File: src/Gaufrette/Adapter/Ftp.php
Comment (muchafm) [inline src/Gaufrette/Adapter/Ftp.php:562]:
```suggestion
     * @param array<int, string>|bool $info
```

File: src/Gaufrette/StreamWrapper.php
Comment (muchafm) [inline src/Gaufrette/StreamWrapper.php:84]: maybe add a bit of explanation for the number 9 ?

**Review by muchafm — CHANGES_REQUESTED** [review]

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AzureBlobStorage.php
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:95]:
```suggestion
    public function deleteContainer(string $containerName, BlobServiceOptions $options = null): void
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:66]:
```suggestion
    public function createContainer(string $containerName, CreateContainerOptions $options = null): void
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:56]:
```suggestion
    public function setCreateContainerOptions(CreateContainerOptions $options): void
```

File: src/Gaufrette/Adapter/Ftp.php
Comment (KevinArtus) [inline src/Gaufrette/Adapter/Ftp.php:452]:
```suggestion
    private function computePath(string $key): string
```

File: src/Gaufrette/Adapter/Ftp.php
Comment (PedroTroller) [reply src/Gaufrette/Adapter/Ftp.php:562]: Event with false... this method is completly fuck-up !

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (PedroTroller) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:317]:
```suggestion   
   /**
     * @param string|resource $content
     */
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Stream/InMemoryBuffer.php
Comment (KevinArtus) [inline src/Gaufrette/Stream/InMemoryBuffer.php:156]:
```suggestion
    public function stat(): array|bool
```

File: src/Gaufrette/Stream/Local.php
Comment (KevinArtus) [inline src/Gaufrette/Stream/Local.php:53]:
```suggestion
    public function read(int $count): string|bool
```

File: src/Gaufrette/Stream/Local.php
Comment (KevinArtus) [inline src/Gaufrette/Stream/Local.php:132]:
```suggestion
    public function stat(): array|bool
```

File: src/Gaufrette/Adapter/AsyncAwsS3.php
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:23]:
```suggestion
```

**Review by muchafm — APPROVED** [review]

**Pull request merged**
